### PR TITLE
Couple of code fixes

### PR DESCRIPTION
**Pull request by jaredpar**
Corrects a couple of items noted in #7677

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-compiler, @dotnet/roslyn-ide 

A code analysis utility was run against the Roslyn repo and filed as #7677.  I did a quick sweep through the issues noted and fixed up a number of them in this PR.  The type of changes fixed were:

- Double writes to a local with no read in the middle.  Typically in unit tests where we forgot to make an assertion on the result of an operation.  
- Incorrect usages of a format string in a `string.Format` call. 
- An `if` statement or `?:` operation where `true` and `false` were the same. 
- Bad `null` checks on the result of `as` expressions

File: src/Compilers/CSharp/Portable/Binder/InMethodBinder.cs [hunk 1]
```diff
@@ -274,13 +274,6 @@ internal static bool ReportConflictWithParameter(Symbol parameter, Symbol newSym
                     return false;
                 }
 
-                if (newSymbolKind == SymbolKind.Parameter || newSymbolKind == SymbolKind.Local)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/InMethodBinder.cs]: Expand the review up a few lines and you'll see an identical check being made which unconditionally returns in the `if` body.  Hence this code path should never get hit.

File: src/Compilers/CSharp/Portable/Binder/WithLambdaParametersBinder.cs [hunk 2]
```diff
@@ -135,13 +135,6 @@ private bool ReportConflictWithParameter(ParameterSymbol parameter, Symbol newSy
                 return true;
             }
 
-            if (newSymbolKind == SymbolKind.Parameter || newSymbolKind == SymbolKind.Local)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/WithLambdaParametersBinder.cs]: Expand the review up a few lines and you'll see an identical check being made which unconditionally returns in the `if` body.  Hence this code path should never get hit.

File: src/Features/Core/Portable/ExtractMethod/AbstractSyntaxTriviaService.Result.cs [hunk 3]
```diff
@@ -113,13 +113,11 @@ private int GetFirstEndOfLineIndex(List<SyntaxTrivia> list)
 
                 // check variable assumption. ordering of two pairs can't be changed
                 Contract.ThrowIfFalse(
-                    (tokens[TriviaLocation.BeforeBeginningOfSpan].RawKind == 0 && tokens[TriviaLocation.AfterEndOfSpan].RawKind == 0) ||
                     (tokens[TriviaLocation.BeforeBeginningOfSpan].RawKind == 0 /* && don't care */) ||
```
Comment (jaredpar) [inline src/Features/Core/Portable/ExtractMethod/AbstractSyntaxTriviaService.Result.cs]: This condition is more inclusive than the deleted one.

File: src/Features/Core/Portable/ExtractMethod/AbstractSyntaxTriviaService.Result.cs [hunk 4]
```diff
@@ -113,13 +113,11 @@ private int GetFirstEndOfLineIndex(List<SyntaxTrivia> list)
...
                     (/* don't care && */ tokens[TriviaLocation.AfterEndOfSpan].RawKind == 0) ||
                     (tokens[TriviaLocation.BeforeBeginningOfSpan].Span.End <= tokens[TriviaLocation.AfterEndOfSpan].SpanStart));
 
                 Contract.ThrowIfFalse(
-                    (tokens[TriviaLocation.AfterBeginningOfSpan].RawKind == 0 && tokens[TriviaLocation.BeforeEndOfSpan].RawKind == 0) ||
                     (tokens[TriviaLocation.AfterBeginningOfSpan].RawKind == 0 /* && don't care */) ||
```
Comment (jaredpar) [inline src/Features/Core/Portable/ExtractMethod/AbstractSyntaxTriviaService.Result.cs]: This condition is more inclusive than the deleted one.

File: src/Compilers/CSharp/Portable/Binder/InMethodBinder.cs (hunk 1 above)
Comment (jasonmalinowski) [inline src/Compilers/CSharp/Portable/Binder/InMethodBinder.cs]: Any sense how this snuck in? Merge issue?

File: src/Workspaces/Core/Portable/Utilities/AbstractSpeculationAnalyzer.cs [hunk 5]
```diff
@@ -380,7 +380,7 @@ private static bool SymbolsAreCompatibleCore(ISymbol symbol, ISymbol newSymbol,
                     var type = methodSymbol.ContainingType;
                     var newType = newMethodSymbol.ContainingType;
                     if ((type != null && type.IsEnumType() &&
-                         type.EnumUnderlyingType != null && type.EnumUnderlyingType.SpecialType == newType.SpecialType) ||
+                         type.EnumUnderlyingType != null && type.EnumUnderlyingType.SpecialType == newType?.SpecialType) ||
```
Comment (jaredpar) [inline src/Workspaces/Core/Portable/Utilities/AbstractSpeculationAnalyzer.cs]: The second part of the overarching `&&` condition indicates that `newType == null` is a real possibility here.  Hence I used an `?.` to account for that.  I'm not particularly familiar with this code hence there may be a better fix here.

File: src/Compilers/CSharp/Portable/Binder/InMethodBinder.cs (hunk 1 above)
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/InMethodBinder.cs]: It was done as a part of a single commit.  It's possible there was a local merge issue that caused this though.

File: src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs [hunk 6]
```diff
@@ -2266,16 +2266,8 @@ public static void Main(string[] args)
                 var thread1 = new Thread(() => comp.GlobalNamespace.GetMember<NamedTypeSymbol>("A").GetMembers());
```
Comment (jasonmalinowski) [inline src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: Is this test actually useful...?

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 7]
```diff
@@ -7191,9 +7192,11 @@ public void NoWarnAndWarnAsError_AnalyzerDriverWarnings()
 
             // TEST: Verify that compiler warning CS8032 can be suppressed via /warn:0.
             output = VerifyOutput(dir, file, additionalFlags: new[] { "/warn:0" });
+            Assert.True(string.IsNullOrEmpty(output));
 
             // TEST: Verify that compiler warning CS8032 can be individually suppressed via /nowarn:.
             output = VerifyOutput(dir, file, additionalFlags: new[] { "/nowarn:CS8032" });
+            Assert.True(string.IsNullOrEmpty(output));
```
Comment (jasonmalinowski) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: Why have the variable? And if null isn't a valid return, why not just Assert.Empty(VerifyOutput(...))?

File: src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs (hunk 6 above)
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: It seems to be forcing us to evaluate the code path in parallel and verify that we updated the result atomically.  It's not perfect but seems to be a healthy excercise of a parallel code path.

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs (hunk 7 above)
Comment (jaredpar) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: All of the tests in this file have this particular pattern.  I tried to follow the existing pattern here vs. establishing a new one.

File: src/Features/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.DiagnosticAsyncToken.cs [hunk 8]
```diff
@@ -29,7 +29,7 @@ protected internal sealed class DiagnosticAsyncToken : AsyncToken
                 int lineNumber)
                 : base(listener)
             {
-                Name = Name;
+                Name = name;
```
Comment (jaredpar) [inline src/Features/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.DiagnosticAsyncToken.cs]: In case you were wondering who could make such a blatant typo ... it was me :frowning:

File: src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs [hunk 9]
```diff
@@ -2266,16 +2266,8 @@ public static void Main(string[] args)
...
-                    thread1.Start();
-                    thread2.Start();
-                }
-                else
-                {
-                    thread1.Start();
-                    thread2.Start();
-                }
+                thread1.Start();
+                thread2.Start();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: The intent here could be to start threads in a different order depending on the condition.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadingAttributes.cs [hunk 10]
```diff
@@ -548,14 +548,14 @@ private void DumpAttributes(Symbol s)
                 int j = 0;
                 foreach (var pa in sa.CommonConstructorArguments)
                 {
-                    Console.WriteLine("{0} {1} {2}", pa.ToString());
+                    Console.WriteLine(pa.ToString());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadingAttributes.cs]: It looks strange for a test to dump something to console. Perhaps, if the output isn't checked in any obvious way, we should simply remove the method.

File: src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs (hunk 9 above)
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: True. I'm happy to change this to do exactly that.

File: src/Compilers/Extension/CompilerPackage.cs [hunk 11]
```diff
@@ -100,9 +95,14 @@ private void WriteMSBuildFiles(string packagePath, string hiveName)
             }
             catch (Exception e)
             {
+                var msg = 
+$@"{e.Message}
+
+To reload the Roslyn compiler package, close Visual Studio and any MSBuild processes, then restart Visual Studio.";
+
```
Comment (AlekseyTs) [inline src/Compilers/Extension/CompilerPackage.cs]: It looks like this message is shown to a user and, therefore, should be localizable. Perhaps we should open a separate issue for this.

File: src/Workspaces/Core/Portable/Utilities/AbstractSpeculationAnalyzer.cs (hunk 5 above)
Comment (AlekseyTs) [inline src/Workspaces/Core/Portable/Utilities/AbstractSpeculationAnalyzer.cs]: It looks like the condition below has the same issue, but with ```type```. Also, it looks like both ```type``` and ```newType``` must be not null in order for the whole condition to be true. Consider checking this fact for both types first and then continue with the ```||``` condition without any null checks for them.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadingAttributes.cs (hunk 10 above)
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadingAttributes.cs]: I agree.  Will delete.

File: src/Compilers/Extension/CompilerPackage.cs (hunk 11 above)
Comment (jaredpar) [inline src/Compilers/Extension/CompilerPackage.cs]: @jasonmalinowski is there a reason this string isn't localized?  If not I'll open a bug.

File: src/Workspaces/Core/Portable/Utilities/AbstractSpeculationAnalyzer.cs (hunk 5 above)
Comment (jaredpar) [inline src/Workspaces/Core/Portable/Utilities/AbstractSpeculationAnalyzer.cs]: That's a good observation.  Will update the check accordingly.

**Comment by jaredpar** [comment]
Updated to respond to PR feedback.

**Comment by gafter** [comment]
LGTM

**Comment by VSadov** [comment]
LGTM

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs [hunk 12]
```diff
@@ -2273,8 +2273,8 @@ public static void Main(string[] args)
                 }
                 else
                 {
-                    thread1.Start();
                     thread2.Start();
+                    thread1.Start();
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: Why?

Comment (gafter) [inline src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: Read the context. The purpose of the enclosing `if` statement is to permute the order in which the threads are started.

Comment (jasonmalinowski) [inline src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: The OS makes no guarantee that the threads run in the order started, so why pretend?

Comment (gafter) [inline src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: Because it helps test for nondeterministic behavior.

**Comment by jaredpar** [comment]
Anyone from @dotnet/roslyn-ide able to sign off on the IDE portion of the change?

File: src/Test/Utilities/Shared/Assert/EqualityUtil`1.cs [hunk 13]
```diff
@@ -161,7 +169,7 @@ private void ObjectEquals3()
             var allValues = _equalityUnits.SelectMany(x => x.AllValues);
             foreach (var value in allValues)
             {
-                Assert.NotEqual((object)42, value);
+                Assert.False(value.Equals((object)42));
```
Comment (jasonmalinowski) [inline src/Test/Utilities/Shared/Assert/EqualityUtil`1.cs]: Why the change? Just to make sure you don't get surprised by xunit behavior?

**Comment by jasonmalinowski** [comment]
:+1:. Also, please either edit the "PR Feedback" commit with a proper description of the change, or just squash it.

File: src/Test/Utilities/Shared/Assert/EqualityUtil`1.cs (hunk 13 above)
Comment (jaredpar) [inline src/Test/Utilities/Shared/Assert/EqualityUtil`1.cs]: @jasonmalinowski the intent of this check was to verify the behavior of `object.Equals(object)`.  It seemed wrong to rely on this being tested as an implementation detail of `Assert.NotEqual` (and in fact it doesn't guarantee this).  Much more direct to just call it directly.

**Comment by jaredpar** [comment]
@jasonmalinowski will do on the squash.  

During the active review of a PR I generally avoid rebasing if possible. This helps keep the comments relevant and seems to generally let things flow better.  Once it gets signed off I squash, push, wait for Jenknis and merge.
